### PR TITLE
Clarify evaluation of arguments to require.

### LINKED ISSUE #10089: Unexpected require argument expression execution
## Description

I would expect `require(condition, revertReason)` construction to not perform `revertReason` expression. For example, it could contain some parsing of previously happened revert reason:

```solidity
(bool success, bytes memory result) = address(target).call(data);
require(success, RevertReasonParser.parse(result, "Call failed: "));
```

Right now `require` construction behavior [described in docs in a wrong way](https://solidity.readthedocs.io/en/latest/control-structures.html#revert) – two options are not equivalent:

telegram-cloud-photo-size-2-5226429365179494300-y

## Environment

- Compiler version: 0.5.x, 0.6.x, 0.7.x

## Steps to Reproduce

1. Try to deploy code with argument `x` equal to `0`:

    ```solidity
    // SPDX-License-Identifier: MIT

    pragma solidity ^0.6.0;


    contract BadRequire {
        constructor(uint256 x) public {
            require(true, f(x));
        }
        
        function f(uint256 x) public pure returns(string memory)  {
            if (x == 0) {
                revert("Ooops");
            }
            return "Some reason";
        }
    }
    ```

2. I expected this tx to succeed, but it will be reverted with reason "Ooops"

### PR DESCRIPTION
**Pull request by chriseth**
Fixes https://github.com/ethereum/solidity/issues/10089

### CODE REVIEW

File: docs/control-structures.rst [hunk 1]
```diff
@@ -615,7 +615,13 @@ The following example shows how to use an error string together with ``revert``
...
     }
 
-The two syntax options are equivalent, it's developer preference which to use.
+If you provide the reason string directly, then the two syntax options are equivalent, it's developer preference which to use.
+
+.. note::
+    The ``require`` function is evaluated just as any other function.
+    This means that all arguments are evaluated before the function itself is executed.
+    In particular, in ``require(condition, f())`` the function ``f`` is executed even if
+    ``condition`` is false.
```
Comment (chriseth) [inline docs/control-structures.rst]:
```suggestion
    ``condition`` is true.
```

File: docs/control-structures.rst [hunk 2]
```diff
@@ -615,7 +615,13 @@ The following example shows how to use an error string together with ``revert``
...
     }
 
-The two syntax options are equivalent, it's developer preference which to use.
+If you provide the reason string directly, then the two syntax options are equivalent, it's developer preference which to use.
+
+.. note::
+    The ``require`` function is evaluated just as any other function.
+    This means that all arguments are evaluated before the function itself is executed.
+    In particular, in ``require(condition, f())`` the function ``f`` is executed even if
+    ``condition`` is true.
```
Comment (bshastry) [inline docs/control-structures.rst:624]: I assume that order of argument evaluation is unspecified. However, the following test leads me to believe it is left to right?

```
contract C {
        uint x;
        function setValue() internal returns (string memory s) {
                x = 1337;
        }

        function f() public returns (uint) {
                require(x != 1337 , setValue());
                return x;
        }
}
// ====
// compileViaYul: also
// ----
// f() -> 1337
```

Comment (chriseth) [reply docs/control-structures.rst:624]: It is mostly left to right, but still unspecified.

Comment (bshastry) [reply docs/control-structures.rst:624]: Should we warn for undefined behavior when a value/reference is both read and written across the arguments of require such as in the test case above?

Comment (chriseth) [reply docs/control-structures.rst:624]: If we do that, then we should do it in all functions.

I think we can expect people to be able to grasp the general concept of side-effects and references.

Comment (bshastry) [reply docs/control-structures.rst:624]: Okay, sounds reasonable to me but still somewhat risky. Perhaps we should add such a test case to document current behavior should order of evaluation change at some point in the future?

File: docs/control-structures.rst [hunk 3]
```diff
@@ -615,7 +615,13 @@ The following example shows how to use an error string together with ``revert``
         }
     }
 
-The two syntax options are equivalent, it's developer preference which to use.
+If you provide the reason string directly, then the two syntax options are equivalent, it's developer preference which to use.
```
Comment (hrkrshnn) [inline docs/control-structures.rst]:
```suggestion
If you provide the reason string directly, then the two syntax options are equivalent, it is the developer's preference which one to use.
```

**Review by hrkrshnn — DISMISSED** [review]

**Comment by hrkrshnn** [comment]
Squashing the commits would be nice.

**Pull request merged**